[offloader] v2: Hide weight onloading latency via prefetching

minosfuture opened the pull request:
<!-- markdownlint-disable -->

## Purpose

This PR adds CPU weight offloader that hides weight onloading latency by prefetching weights.  This saves the performance cost of zero-copy UVA access. This technique was first developed in SGLang for GB200: https://lmsys.org/blog/2025-09-25-gb200-part-2/, and now adapted to support torch.compile and CUDA graph within vLLM in this PR.

Also refactors the offloading to be extensible. 

Demonstrated in the trace:
* H2D is for prefetching next offloaded weights

<img width="2318" height="126" alt="image" src="https://github.com/user-attachments/assets/0bd4202f-7797-4f87-b604-5da64503d102" />

## Test Plan

* deepseek fp4 in 2 GPUs with offloading
* deepseek fp4 in 4 GPUs without offloading (doesn't fit in 2 GPUs without offloading)

Example serving recipe for GB200:

```
HF_HUB_OFFLINE=1 NVIDIA_GDRCOPY=1 NVSHMEM_IB_ENABLE_IBGDA=1 VLLM_SKIP_P2P_CHECK=1 \
NCCL_CUMEM_ENABLE=1 NCCL_MNNVL_ENABLE=1 NCCL_NVLS_ENABLE=1 \
VLLM_RANDOMIZE_DP_DUMMY_INPUTS=1 VLLM_USE_FLASHINFER_MOE_FP4=1 \
VLLM_FLASHINFER_MOE_BACKEND=latency VLLM_ENABLE_FUSED_MOE_ACTIVATION_CHUNKING=0 \
VLLM_USE_NCCL_SYMM_MEM=1 VLLM_ENABLE_MOE_DP_CHUNK=0 numactl --cpunodebind=0 \
--membind=0 vllm serve \
nvidia/DeepSeek-R1-0528-NVFP4-v2 \
--attention-config.backend FLASHINFER_MLA \
--attention-config.use_trtllm_ragged_deepseek_prefill=true --kv-cache-dtype fp8 \
--enable-expert-parallel --data-parallel-rpc-port 13345 --max-model-len 2148 \
--disable-uvicorn-access-log --port 8000 --trust-remote-code --async-scheduling \
--disable_custom_all_reduce --disable_nccl_for_dp_synchronization \
--no-enable-prefix-caching --all2all-backend allgather_reducescatter \
--gpu-memory-utilization 0.85 --max-num-batched-tokens 65536 --max-num-seqs 1024 \
--swap-space 16 --data-parallel-size 2 --offload-group-size 2 \
--offload-num-in-group 1 --offload-prefetch-step 1 --offload-params w13_weight \
w2_weight --api-server-count 1 --profiler-config '{"profiler": "torch",
"torch_profiler_dir": "/tmp/traces"}' 2>&1 | tee /tmp/dsr1.log
```

## Test Result


### QPS per GPU 18.04 (2 GPUs)

this PR `--offload-group-size 2 --offload-num-in-group 1 --offload-prefetch-step 1`

```
============ Serving Benchmark Result ============
Successful requests:                     1024
Failed requests:                         0
Maximum request concurrency:             1024
Benchmark duration (s):                  56.78
Total input tokens:                      2096128
Total generated tokens:                  1024
Request throughput (req/s):              18.04
Output token throughput (tok/s):         18.04
Peak output token throughput (tok/s):    63.00
Peak concurrent requests:                1024.00
Total Token throughput (tok/s):          36937.17
---------------Time to First Token----------------
Mean TTFT (ms):                          30965.61
Median TTFT (ms):                        30190.67
P99 TTFT (ms):                           56685.92
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00
Median TPOT (ms):                        0.00
P99 TPOT (ms):                           0.00
---------------Inter-token Latency----------------
Mean ITL (ms):                           0.00
Median ITL (ms):                         0.00
P99 ITL (ms):                            0.00
==================================================
```


### QPS per GPU 16.9 (4 GPUs)

without `--offload-group-size 2 --offload-num-in-group 1 --offload-prefetch-step 1`

```
============ Serving Benchmark Result ============
Successful requests:                     1024
Failed requests:                         0
Maximum request concurrency:             1024
Benchmark duration (s):                  30.28
Total input tokens:                      2096128
Total generated tokens:                  1024
Request throughput (req/s):              33.82
Output token throughput (tok/s):         33.82
Peak output token throughput (tok/s):    64.00
Peak concurrent requests:                1024.00
Total Token throughput (tok/s):          69259.31
---------------Time to First Token----------------
Mean TTFT (ms):                          17335.01
Median TTFT (ms):                        17177.41
P99 TTFT (ms):                           30175.24
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00
Median TPOT (ms):                        0.00
P99 TPOT (ms):                           0.00
---------------Inter-token Latency----------------
Mean ITL (ms):                           0.00
Median ITL (ms):                         0.00
P99 ITL (ms):                            0.00
==================================================
```

### Accuracy:

local-completions (model=nvidia/DeepSeek-R1-0528-FP4-v2,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=32), gen_kwargs: (None), limit: 100.0, num_fewshot: None, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.98|±  |0.0141|
|     |       |strict-match    |     5|exact_match|↑  | 0.98|±  |0.0141|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

minosfuture committed:
offloader v2

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
cleanup

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
cleanup: extract func; move imports

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
fix import

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

minosfuture committed:
address comment: remove legacy code

Signed-off-by: Ming Yang <minos.future@gmail.com>

ywang96 reviewed (commented):
IIUC this is mostly a RL feature, correct? Maybe @youkaichao can take a look?

elvircrn commented:
Why is TPOT and ITL 0 in the benchmark result?

mgoin commented on vllm/model_executor/models/deepseek_v2.py:
```diff
@@ -1275,6 +1275,31 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+                    layer.mlp.experts
+                    if isinstance(layer.mlp, DeepseekV2MoE)
+                    else layer.mlp
+                ),
```
Maybe we could generalize this, at least for the moe, by searching for an attr that isinstance of FusedMoE

mgoin commented on vllm/model_executor/models/deepseek_v2.py:
```diff
@@ -1275,6 +1275,31 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+                            ["w13_blockscale_swizzled", "w2_blockscale_swizzled"]
+                            if hasattr(module, "w13_blockscale_swizzled")
+                            else []
+                        ),
```
Is there harm in offloading "too many" params? I feel like I'd prefer a general "w13" or "w2" match rather than specializing on specific quant methods

mgoin commented on vllm/config/cache.py:
```diff
@@ -89,6 +89,16 @@ class CacheConfig:
+    Example: group_size=8, num_in_group=2 offloads layers 6,7,14,15,22,23,...
+    """
+    offload_num_in_group: int = Field(default=1, ge=1)
+    """Advanced CPU offloading: Number of layers to offload per group. Default is 1."""
```
I didn't see any validation that that offload_num_in_group <= offload_group_size

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

minosfuture committed:
refactor: move offload config from CacheConfig to OffloadConfig

Address review comments:
1. Add validation that offload_num_in_group <= offload_group_size
2. Move offload settings out of CacheConfig (not KV cache related)

Changes:
- Create new OffloadConfig in vllm/config/offload.py with all
  model weight offloading settings (cpu_offload_gb, offload_group_size,
  offload_num_in_group, offload_prefetch_step)
- Add pydantic validator to ensure offload_num_in_group <= offload_group_size
- Add offload_config field to VllmConfig
- Update arg_utils.py to create OffloadConfig and use separate arg group
- Update offloader/base.py to accept OffloadConfig instead of CacheConfig
- Update gpu_model_runner.py to use offload_config
- Mark cpu_offload_gb in CacheConfig as deprecated (kept for backward compat)

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
refactor: generalize MoE detection for offloading

Address review comment about generalizing MoE detection by searching
for FusedMoE instances instead of hardcoding specific model classes.

Changes:
- Add find_fused_moe_submodule() helper to fused_moe/layer.py
- Export the new helper from fused_moe/__init__.py
- Replace hardcoded DeepseekV2MoE check with generic helper

This makes the offloader configuration reusable across all MoE models
(Qwen2MoE, Mixtral, etc.) without model-specific imports.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
cleanup: remove NVFP4 scale params from offload whitelist

Remove w13_blockscale_swizzled and w2_blockscale_swizzled from the
offload parameter whitelist - only offload the core expert weights.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
cleanup: remove dead code in UVAOffloader

Remove redundant uva_offloading variable and dead code branches.
Since UVA is required (asserted in __init__), the non-UVA code paths
were never executed.

Signed-off-by: Ming Yang <minos.future@gmail.com>

youkaichao reviewed (commented):
mainly thinking about how to make it work for `torch.compile` + cudagraph.

I think it's possible to have a static buffer `workspace` (rough size `prefetch_layers * params_offloaded_per_layer)`, and then after offloading:

```python
cpu_tensor = w
w = workspace[start:end].reshape(xxx)
w.cpu_tensor = cpu_tensor
```

then we insert one pytorch custom op `prefetch(layer_name)` in every layer that contains offloaded parameters, which copies cpu weights to the slice of the workspace buffer. it seems to me this can be compatible with `torch.compile`. However, it might be dangerous that `torch.compile` might rearrange the op while it actually inplace mutates many weight tensors. cc @zou3519

youkaichao commented:
> However, it might be dangerous that torch.compile might rearrange the op while it actually inplace mutates many weight tensors.

traditional compilers have the concept of "volatile" that prevents the compiler doing certain optimizations. Not sure if `torch.compile` has similar tools.

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[Core] V2 offloader: static buffers and custom ops for torch.compile

This commit introduces V2 offloader improvements for torch.compile and
CUDA graph compatibility:

1. Static GPU buffer pool with double/triple buffering support
2. Custom ops (wait_prefetch, start_prefetch) that create data dependencies
   via mutates_args to prevent torch.compile from reordering operations
3. Stream synchronization to ensure correct prefetch/compute ordering
4. Stride-preserving buffers to maintain parameter memory layout

Key design decisions:
- Parameters point to static GPU buffers (not CPU storage) so torch.compile
  sees GPU tensors during tracing
- sync_cpu_storage() updates CPU storage after process_weights_after_loading
  to capture quantization-processed weights
- Buffer slot keys include parameter name to prevent different parameters
  with same shape from sharing buffers within a layer

Verified with lm_eval gsm8k accuracy >0.94 on DeepSeek-R1-0528-NVFP4-v2.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[Core] V2 offloader: CUDA graph capture support

Add CUDA graph capture compatibility for the V2 offloader when used
with torch.compile. This enables CPU weight offloading to work with
piecewise and full cudagraph modes.

Problem:
When CUDA graph capture started, pre-capture prefetches on copy_stream
caused errors:
- cudaErrorStreamCaptureIsolation: dependency on uncaptured work
- cudaErrorStreamCaptureUnjoined: forked stream not rejoined
- cudaErrorInvalidValue: waiting on invalid event state

Solution:
Use event-based stream forking during capture and wait_stream outside
capture. Key mechanisms:

1. Event-based fork/join during capture:
   - Record fork_event on compute stream before prefetch
   - copy_stream.wait_event(fork_event) joins it to capture
   - Record _copy_done_event after H2D copies complete
   - compute_stream.wait_event(_copy_done_event) creates dependency

2. Conditional synchronization:
   - _prefetch_in_capture flag tracks if prefetch was during capture
   - During capture: skip wait_event for pre-capture prefetches
   - Outside capture: use wait_stream for robustness

3. Capture boundary handling:
   - sync_before_graph_capture(): ensure pre-capture work complete
   - join_after_forward(): rejoin copy_stream after model forward
   - Reset flags between piecewise cudagraph subgraphs

Integration points:
- cuda_graph.py: CUDAGraphWrapper capture
- cudagraph_utils.py: V1 worker capture
- eagle_cudagraph.py: speculative decoding capture
- gpu_ubatch_wrapper.py: micro-batch wrapper capture

With this fix, relaxed capture mode (capture_error_mode="relaxed") is
no longer needed and has been removed from all capture sites.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
clean up unnecessary code

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[Core] V2 offloader: simplify API by using get_offloader() directly

Remove wrapper functions from v2_ops.py and add methods to
BaseOffloader so call sites use get_offloader() directly.

Changes:
- Add sync_prev_onload(), join_after_forward(), _wait_for_layer(),
  _start_prefetch() to BaseOffloader with no-op defaults
- Remove sync_offloader_before_capture() and join_offloader_after_forward()
  wrapper functions from v2_ops.py
- Update cuda_graph.py, cudagraph_utils.py, gpu_ubatch_wrapper.py,
  eagle_cudagraph.py to call get_offloader() methods directly
- Rename sync_before_graph_capture to sync_prev_onload (more accurate
  since it's called before both capture and replay)

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[Core] V2 offloader: fix CUDA graph bugs in Eagle and UBatch

Fix three bugs in V2 offloader CUDA graph support:

1. EagleCudaGraphManager: add sync_prev_onload() before capture to
   ensure warmup prefetches complete before graph capture starts

2. UBatchWrapper: add join_after_forward() inside capture block to
   avoid unjoined stream error from last layer's start_prefetch

3. Config validation: require offload_prefetch_step >= 1 when V2
   offloading is enabled to prevent ZeroDivisionError in StaticBufferPool

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

minosfuture committed:
minor param order adjustmetn

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Keep cpu_offload_params on CacheConfig as deprecated field

Adding back cpu_offload_params to CacheConfig to avoid breaking
downstream code that constructs CacheConfig(cpu_offload_params=...)
directly. The field defaults to an empty set and directs users to
OffloadConfig.cpu_offload_params.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Fix OffloadConfig.compute_hash() returning constant hash

OffloaderV2 patches module forwards and inserts custom ops
(wait_prefetch, start_prefetch) into the computation graph,
so offload settings must be part of the compilation cache key.
The previous constant hash caused stale cache hits when switching
between offloaded and non-offloaded runs of the same model.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Fix OffloaderV2 accuracy bug: re-pin CPU storage after process_weights_after_loading

After process_weights_after_loading, device_loading_context creates
non-pinned CPU tensors via `p.data = p.data.to("cpu")`. When
sync_cpu_storage() or assign_static_buffer() adopted these non-pinned
tensors as _cpu_storage, subsequent non_blocking=True H2D copies would
trigger an implicit CUDA stream synchronization (cudaMemcpyAsync from
pageable memory). This broke the event-based fork synchronization
between the compute and copy streams, allowing the copy to overwrite
the GPU buffer while the compute stream was still reading from it.

Fix by extracting _update_cpu_storage_from_param() that re-pins CPU
tensors when needed, and adding an assertion in start_onload_to_static
to catch non-pinned CPU storage early.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
pre-commit fixes

Signed-off-by: Ming Yang <minos.future@gmail.com>

zou3519 commented on vllm/model_executor/offloader/v2_ops.py:
```diff
@@ -0,0 +1,102 @@
+    return output_tensor
+
+
+def register_v2_offloader_ops() -> None:
```
I think I see what's going on, I'm wrong about the CPU->GPU transfers. we can indeed cudagraph-capture the H2D as long as the addresses are static

minosfuture committed:
Fix OffloaderV2 crash with dotted parameter names

_BaseParamOffloader._param used plain getattr() which doesn't
traverse dotted paths like 'self_attn.qkv_proj.weight'. Models
that don't provide a whitelist_param_names_creator (e.g. Llama)
hit this because named_parameters() yields fully-qualified names.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Specify removal version for deprecated CacheConfig offload fields

Per vllm deprecation policy, deprecated parameters must specify which
version they will be removed in. Updated cpu_offload_gb and
cpu_offload_params on CacheConfig to target v0.16 for removal.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Use @config decorator alone for OffloadConfig

The @config decorator already applies pydantic dataclass internally,
so the explicit @dataclass was redundant. Aligns with how other config
classes (e.g. CacheConfig) are defined.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Add explicit offload_backend selector with nested sub-configs

Introduce OffloadBackend ("auto"|"uva"|"prefetch"), UVAOffloadConfig,
and PrefetchOffloadConfig as nested sub-configs of OffloadConfig.
The --offload-backend CLI flag lets users explicitly choose a backend
instead of relying on implicit field-based detection.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Rename OffloaderV2 to PrefetchOffloader

Rephrase "V2 offloading" as "prefetching-based offloading" throughout.
The two backends (UVA and prefetch) serve different use cases rather
than being iterations of the same approach.

Renames:
- OffloaderV2 -> PrefetchOffloader
- v2.py -> prefetch.py
- v2_ops.py -> prefetch_ops.py
- test_v2_offload.py -> test_prefetch_offload.py

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Move PrefetchOffloader parameterization from model code to CLI

Add --offload-params CLI flag to PrefetchOffloadConfig so parameter
selection is driven entirely from config, matching how UVA's
--cpu-offload-params works. This removes model-specific Python
callbacks (submodule_accessor, whitelist_param_names_creator) from
the offloader interface, making prefetch offloading work
out-of-the-box for all models without per-model code changes.

- Add offload_params field to PrefetchOffloadConfig with segment matching
- Add --offload-params CLI flag wired through EngineArgs
- Replace callback-based wrap_modules with config-driven matching
- Simplify BaseOffloader/NoopOffloader/UVAOffloader signatures
- Remove offloader_kwargs from make_layers() and deepseek_v2.py
- Delete find_fused_moe_submodule (no longer needed)
- Add --offload-params to prefetch offload test

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Make prefetch custom ops return None instead of tensor

mutates_args already creates data dependencies for torch.compile,
so returning the tensor is unnecessary. Use in-place mutation
semantics and simplify the call sites in _hook_module_forward.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Switch prefetch offload test from DeepSeek-V2-Lite to Llama-3.2-1B-Instruct

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

# Conflicts:
#	vllm/v1/worker/gpu/cudagraph_utils.py
#	vllm/v1/worker/gpu/spec_decode/eagle/cudagraph.py

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture commented:
@mgoin Hi Michael, could you help review for another round? I hope I've addressed your earlier concerns by moving param config to cli so we can run this with any models now. cc @tlrmchlsmth

minosfuture commented on vllm/model_executor/offloader/v2_ops.py:
```diff
@@ -0,0 +1,102 @@
+    layer_idx: int,
+) -> torch.Tensor:
+    """Fake implementation for torch.compile tracing."""
+    return output_tensor
```
updated!

minosfuture committed:
Remove incorrect comment

Signed-off-by: Ming Yang <minos.future@gmail.com>

eellison commented on vllm/model_executor/offloader/prefetch.py:
```diff
@@ -0,0 +1,691 @@
+            # Wait for this layer's prefetch to complete
+            # mutates_args on input_tensor creates data dependency for torch.compile
+            input_tensor = args[0] if args else kwargs.get("hidden_states")
+            torch.ops.vllm.wait_prefetch(input_tensor, index)
```
cc @zou3519 , will this result in an unnecessary clone at runtime? do we need an alternative api to declare dependencies ? e.g. effects / control deps etc ?

eellison reviewed (commented):
couple comments

minosfuture committed:
Use per-layer event sync in eager mode to enable H2D overlap

In eager mode, _wait_for_layer was using wait_stream(copy_stream)
which waits for ALL pending H2D copies, serializing every layer's
prefetch. Switch to per-layer wait_event so each layer only waits
for its own copy, allowing prefetch_step>1 to overlap H2D with
compute from intervening layers.

Events recorded during cudagraph capture become invalid after
capture ends, so we track _event_valid_for_eager and fall back
to wait_stream when the event was last recorded in a capture
context.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture commented on vllm/model_executor/offloader/prefetch.py:
```diff
@@ -0,0 +1,691 @@
+            param_infos=param_infos,
+            slot_capacity=self.prefetch_step,
+            device=device,
+        )
```
prefetch starts at the end of current layer, illustrated in the figure below to compare prefetch_step==1  vs. prefetch_step==2 (double buffering).
<img width="1952" height="1070" alt="image" src="https://github.com/user-attachments/assets/c55a26c2-bb87-4737-a9a9-7ab900ef656e" />

minosfuture committed:
Add offload_params to LLM interface

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Use torch.finfo instead of creating tensor to get element size

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Initialize global offloader as None to fail loudly before set_offloader

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Add nightly e2e test for prefetch offloading with DeepSeek-V2-Lite

Runs DeepSeek-V2-Lite with prefetch offloading of MoE expert weights
(w13_weight, w2_weight) and validates GSM8K accuracy matches baseline.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Fix sync_prev_onload comments in run_fullgraph

Signed-off-by: Ming Yang <minos.future@gmail.com>

mgoin committed:
Update .buildkite/test_areas/e2e_integration.yaml

Signed-off-by: Michael Goin <mgoin64@gmail.com>

mgoin reviewed (approved):
LGTM, thanks for the changes. I think it is obvious the `get_offloader().sync_prev_onload()` and `get_offloader().join_after_forward()` insertions in the model runner are fragile and prone to failure in the future, but I'm not sure how to better structure this. Maybe @benchislett or @LucasWilkinson have strong opinions against this, but we do need this performant feature to land sooner or later

minosfuture committed:
Default global offloader to NoopOffloader and log on set

Set _instance default to NoopOffloader() so get_offloader() always
returns a valid instance. Log the offloader type in set_offloader()
for visibility into which backend is active.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
Merge remote-tracking branch 'minosfuture/offloader' into offloader

minosfuture committed:
Merge remote-tracking branch 'origin/main' into offloader

mgoin committed:
Merge branch 'main' into offloader

zou3519 reviewed (approved):
compile pieces lgtm
